File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* ans = nullptr;
    ListNode* helper(ListNode* node) {
        if(node == nullptr || node->next == nullptr) {
            ans = node;
            return node;
        }
        ListNode* curr = helper(node->next);
        curr->next = node;
        node->next = nullptr;
        return node;
    }
    ListNode* reverseList(ListNode* head) {
        helper(head);
        return ans;
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    ans = None
    def helper(self, node: Optional[ListNode]) -> Optional[ListNode]:
        if node == None or node.next == None:
            self.ans = node
            return node
        curr = self.helper(node.next)
        curr.next = node
        node.next = None
        return node
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        self.helper(head)
        return self.ans